`tf2_geometry_msgs`: Fixing covariance transformation in `doTransform<PoseWithCovarianceStamped, TransformStamped>`

aprotyas opened the pull request:
Previously, the covariance matrix of a `PoseWithCovarianceStamped` message [was incorrectly transformed](https://github.com/ros2/geometry2/blob/d65b7eb996dcc81bfd3389153b25b9d56eed0e8c/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h#L309). This PR:

- Adds a `transformCovariance` method to address the covariance transformation.
- Updates the `TfGeometry/FrameWithCovariance` unit test to check for the expected transformed covariance.

Fixes #372

Commit:
Implement covariance transformation (cpp)

* The transform operation has been performed with the row-major form
of the covariance matrix.

* tf2 matrix types were used for easier matrix multiplication.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

Commit:
Fixing `TfGeometry/FrameWithCovariance` test

* The previous test was doing a wrong covariance check,
i.e. out.covariance = in.covariance.

* The updated test checks against the correctly transformed
covariance matrix.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

sloretz commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp:
```diff
@@ -347,6 +349,104 @@ template <>
+
+    // get rotation matrix (and transpose)
+    tf2::Transform tf_transform;
+    fromMsg(transform, tf_transform);
```
Say someone already has a transform as type `tf2::Transform`, and they want to use `transformCovariance()`. They would need to convert it to `geometry_msgs::msg::Transform`, but then this function would internally convert it back. Both of those conversion steps would have been unnecessary.

Since the function operates on the `tf2::Transform` type, I would recommend making that the type of the argument. The caller's can handle transforming it from whatever datatype they have.

aprotyas replied:
If I understand correctly, you are suggesting changing the `geometry_msgs::msg::Transform` argument to `tf2::Transform`, right? I debated doing that, but I wasn't sure what would be the more common usage. Given `tf2::Transform` is the more general "transform", and not just a message type, I think I will change that argument type.

aprotyas replied:
Addressed in 8545842360dad058

sloretz commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp:
```diff
@@ -347,6 +349,104 @@ template <>
+     * To transform a covariance matrix:
+     * 
+     * \verbatim[R 0] COVARIANCE [R' 0 ]
+      [0 R]            [0  R']\endverbatim
```
I've never used the `\verbatim` command before. How does it look when the documentation is generated?

aprotyas replied:
I can't include a picture here, but the documentation looks like:
```
[R 0] COVARIANCE [R' 0 ]
[0 R]            [0  R']
```

sloretz commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp:
```diff
@@ -347,6 +349,104 @@ template <>
+ * \return The transformed covariance matrix.
+ */
+inline
+geometry_msgs::msg::PoseWithCovariance::_covariance_type transformCovariance(const geometry_msgs::msg::PoseWithCovariance::_covariance_type& cov_in, const geometry_msgs::msg::Transform& transform)
```
It looks like this file isn't being linted, but I'd still recommend [limiting the line length to 100 characters](https://docs.ros.org/en/rolling/Contributing/Code-Style-Language-Versions.html#line-length) for readability.

aprotyas replied:
This is what `ament_uncrustify` recommends as well, so I will do this.

```suggestion
geometry_msgs::msg::PoseWithCovariance::_covariance_type transformCovariance(
  const geometry_msgs::msg::PoseWithCovariance::_covariance_type & cov_in,
  const geometry_msgs::msg::Transform & transform)
```

aprotyas replied:
Addressed in 8545842360dad058

sloretz commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp:
```diff
@@ -347,6 +349,104 @@ template <>
+			         cov_in[33], cov_in[34], cov_in[35]);
+
+    // perform blockwise matrix multiplication
+    const tf2::Matrix3x3 result_11 = R * cov_11 * R.transpose();
```
`R.transpose()` returns [a new `Matrix3x3` instance](https://github.com/ros2/geometry2/blob/6fa2001c026ff104b890663b52f2e3d518140146/tf2/include/tf2/LinearMath/Matrix3x3.h#L557-L558), so it may be more efficient to store it in a local variable instead of calling `R.transpose()` multiple times.

This comment is premature optimization so feel free to ignore.

aprotyas replied:
Addressed in 8545842360dad058

sloretz reviewed (commented):
The math LGTM (I only checked that it matches [the ROS 1 code](https://github.com/ros/geometry2/blob/c73b5939723db078c9bbe18523230ad54f859682/tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.h#L927))

aprotyas commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp:
```diff
@@ -367,7 +467,7 @@ inline
     t_out.header.stamp = transform.header.stamp;
     t_out.header.frame_id = transform.header.frame_id;
-    t_out.pose.covariance = t_in.pose.covariance;
+    t_out.pose.covariance = transformCovariance(t_in.pose.covariance, transform.transform);
```
> The caller's can handle transforming it from whatever datatype they have.

You mean something akin to this, right? @sloretz 

```suggestion
tf2::Transform tf_transform;
fromMsg(transform.transform, tf_transform);
t_out.pose.covariance = transformCovariance(t_in.pose.covariance, tf_transform);
```

sloretz replied:
Yup, that's what I mean!

Commit:
Covariance transformation now accepts tf Transform

Previously, `transformCovariance`'s transform argument was of type
`geometry_msgs::msg::Transform`, which is not a "general" transform
type, but rather a message type. As such, that argument type has been
changed to `tf2::Transform`, so that the user can convert from whatever
transform type they have to a tf Transform.

Also:

* Moved location of a forward declaration to be closer to its use.
* Applied `ament_uncrustify` changes for maximum line length, changing
  the `transformCovariance` signature ine.
* Reflected change in `transformCovariance` function signature in
  doxygen documentation.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

aprotyas commented:
CI (build: `--packages-up-to tf2_geometry_msgs` test: `--packages-select tf2_geometry_msgs`)

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=14739)](http://ci.ros2.org/job/ci_linux/14739/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=9490)](http://ci.ros2.org/job/ci_linux-aarch64/9490/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=12404)](http://ci.ros2.org/job/ci_osx/12404/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=14887)](http://ci.ros2.org/job/ci_windows/14887/)

clalancette commented on tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp:
```diff
@@ -347,6 +349,105 @@ template <>
+
+    // convert covariance matrix into four 3x3 blocks
+    const tf2::Matrix3x3 cov_11(cov_in[0], cov_in[1], cov_in[2],
+			         cov_in[6], cov_in[7], cov_in[8],
+			         cov_in[12], cov_in[13], cov_in[14]);
```
It looks like there are leading tabs in these two lines.  It is causing the function parameters to not line up (both in the GitHub UI and in `vim`).  Please convert all tabs to spaces.

The same goes several times below.

aprotyas replied:
Done, thanks to `:retab`.

clalancette reviewed (changes requested):
One minor thing to fix, then this looks good to me.

Commit:
Converting tabs to spaces in `transformCovariance`

Leading tabs prevented the function parameters from lining up, both in
the Github UI and in Vim. Converting all tabs to spaces fixes said
issue.

Signed-off-by: Abrar Rahman Protyasha <abrar@openrobotics.org>

clalancette reviewed (approved):
Looks good to me!

aprotyas commented:
@Mergifyio backport galactic foxy
